atomic_reactor/utils/resource_manager.py

```diff
@@ -6,193 +6,290 @@
+            logger.warning("SSH Connection error. %s", ex)
+        except paramiko.ssh_exception.SSHException as ex:
+            logger.warning("SSH Connection error. %s", ex)
+        return stdout, stderr, client.recv_exit_status()
+
+    def _ssh_run_remote_cmd_to_get_json(self, client, cmd: List):
+        stdin, stdout, stderr = self._ssh_run_remote_cmd(client, cmd)
+        try:
+            json_stdout = json.loads(stdout)
+        except ValueError as ex:  # Not valid JSON  #  TODO: Question - Do we want to fail build on loading invalid json from remote host????
+            logger.warning("Json load failed. Not a valid JSON. %s", ex)
+            return None
+        except TypeError as ex:  # Not an object
+            logger.warning("Json load failed. Not an object. %s", ex)
+            return None
+        except TypeError as ex:  # Unable to decode the bytes
+            logger.warning("Json load failed. Unable to decode the bytes. %s", ex)
+            return None
+        return json_stdout
+
+    def available_slots(self) -> list[int]:
+        """
+        Returns list of available slots
+
+        Slots should be randomized here, for smoother locking, because there will be
+        hunger games when every pipeline will try to lock slot with lowest number first
+        """
+        logger.debug("Retrieve list of available slots on host %s", self._hostname)
+
+        with self._ssh_session() as client:
+            cmd = self.get_cmd_to_read_from_json(RESOURCE_INFO_JSON)
+            resource_info = self._ssh_run_remote_cmd_to_get_json(client, cmd)
+
+            available_slots = []
+            for slot_id in range(resource_info.get('max_slot_count', 1)):
+                cmd = self.get_cmd_to_read_from_json(f'slot_{slot_id}.json')
+                slot = self._ssh_run_remote_cmd_to_get_json(client, cmd)
+
+                if slot is None:  # Json load failed - file on remote host might be corrupted
+                    logger.warning("%s - slot %s - json load failed", resource_info['hostname'], slot_id)
+                elif slot:  # Non-empty json file was loaded
+                    prid = slot.get('prid')
+                    logger.debug("%s - slot %s is occupied by %s", resource_info['hostname'], slot_id, prid)
+                    self._validate_resource_lock(slot)
+                elif not slot:  # Empty json file was loaded
+                    logger.debug("%s - slot %s is available", resource_info['hostname'], slot_id)
+                    available_slots.append(slot_id)
+                else:  # Undefined state - should not happen
+                    raise RuntimeError(f"{resource_info['hostname']} - slot {slot_id} invalid state!")
+
+        return random.shuffle(available_slots)
+
+    def get_cmd_to_read_from_json(self, file_name: str):
+        """ Create a command to retrieve content of json file from remote host """
+        return ['cat', os.path.join(RELATIVE_PATH, file_name)]
```

Comment:
if you define RELATIVE_PATH as pathobject in module, you can just use
```python
str(RELATIVE_PATH / filename)
```

not sure if that's better

Reply:
If there won't be any strong opinion on this, It will stay as it is now. :)